### PR TITLE
Default collection install to first path in COLLECTIONS_PATHS

### LINKED ISSUE #62869: Missing default collection install path
##### SUMMARY
`ansible-galaxy collection install` currently requires the user to specify an install path using the `-p` option. Galaxy currently generates and displays the CLI command for installing a collection. There is no way for Galaxy to determine a value for `-p`, and hard coding a default value into Galaxy makes no sense.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
`ansible-galaxy`

##### ANSIBLE VERSION

```
ansible 2.10.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/Users/chouseknecht/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/chouseknecht/projects/ansible/lib/ansible
  executable location = /Users/chouseknecht/.pyenv/versions/venv27/bin/ansible
  python version = 2.7.14 (default, Nov 14 2017, 23:24:24) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.38)]
```

### PR DESCRIPTION
**Pull request by chouseknecht**
##### SUMMARY
Provide a sane default install path value for `ansible-galaxy collection install`

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
`ansible-galaxy`

##### ADDITIONAL INFORMATION

Fixes #62869

### CODE REVIEW

File: docs/docsite/rst/dev_guide/developing_collections.rst [hunk 1]
```diff
@@ -238,15 +238,19 @@ as a distribution method, but you can use it directly to install the collection
...
 .. code-block:: bash
 
-   ansible-galaxy collection install my_namespace-my_collection-1.0.0.tar.gz -p ./collections/ansible_collections
+   ansible-galaxy collection install my_namespace-my_collection-1.0.0.tar.gz -p ./collections
 
-You should use one of the values configured in :ref:`COLLECTIONS_PATHS` for your path. This is also where Ansible itself will expect to find collections when attempting to use them.
 
-Then try to use the local collection inside a playbook, for more details see :ref:`Using collections <using_collections>`
+You should use one of the values configured in :ref:`COLLECTIONS_PATHS` for your path. This is also where Ansible itself will
+expect to find collections when attempting to use them. If a path value is not specified, collection will be installed to the first path
```
Comment (samccann) [inline docs/docsite/rst/dev_guide/developing_collections.rst]:
```suggestion
expect to find collections when attempting to use them. If you don't specify a path value, ``ansible-galaxy collection`` installs the collection in the first path
```

File: docs/docsite/rst/dev_guide/developing_collections.rst [hunk 2]
```diff
@@ -238,15 +238,19 @@ as a distribution method, but you can use it directly to install the collection
...
+defined in :ref:`COLLECTIONS_PATHS`, which by default is ``~/.ansible/collections``.
+
+Next, try using the local collection inside a playbook, for examples and more details see :ref:`Using collections <using_collections>`
```
Comment (samccann) [inline docs/docsite/rst/dev_guide/developing_collections.rst]:
```suggestion
Next, try using the local collection inside a playbook. For examples and more details see :ref:`Using collections <using_collections>`.
```

File: docs/docsite/rst/user_guide/collections_using.rst [hunk 3]
```diff
@@ -36,7 +36,9 @@ You can also directly use the tarball from your build:
     parent directory is already in a folder called ``ansible_collections``.
 
 
-You should use one of the values configured in :ref:`COLLECTIONS_PATHS` for your path. This is also where Ansible itself will expect to find collections when attempting to use them.
+When specifying the install path, use one of the values configured in :ref:`COLLECTIONS_PATHS`, as this is where Ansible itself will expect to
```
Comment (samccann) [inline docs/docsite/rst/user_guide/collections_using.rst]:
```suggestion
When using the ``-p`` option to specify the install path, use one of the values configured in :ref:`COLLECTIONS_PATHS`, as this is where Ansible itself will expect to
```

File: docs/docsite/rst/user_guide/collections_using.rst [hunk 4]
```diff
@@ -36,7 +36,9 @@ You can also directly use the tarball from your build:
...
+find collections. If no path is specified, a collection will be installed to the first path defined in :ref:`COLLECTIONS_PATHS`, which by
```
Comment (samccann) [inline docs/docsite/rst/user_guide/collections_using.rst]:
```suggestion
find collections. If you don't specify a path, ``ansible-galaxy collection`` installs the collection to the first path defined in :ref:`COLLECTIONS_PATHS`, which by
```

File: docs/docsite/rst/dev_guide/developing_collections.rst (hunk 1 above)
Comment (chouseknecht) [reply docs/docsite/rst/dev_guide/developing_collections.rst]: Fixed

File: docs/docsite/rst/dev_guide/developing_collections.rst (hunk 2 above)
Comment (chouseknecht) [reply docs/docsite/rst/dev_guide/developing_collections.rst]: Fixed

File: docs/docsite/rst/user_guide/collections_using.rst (hunk 3 above)
Comment (chouseknecht) [reply docs/docsite/rst/user_guide/collections_using.rst]: Fixed

File: docs/docsite/rst/user_guide/collections_using.rst (hunk 4 above)
Comment (chouseknecht) [reply docs/docsite/rst/user_guide/collections_using.rst]: Fixed

**Review by samccann — APPROVED** [review]
Docs portion LGTM

**Comment by alikins** [comment]
Sounds correct to me. And in line with the original intentional for 'ansible-galaxy collection install'.

The default for `C.COLLECTIONS_PATH` is `~/.ansible/collections:/usr/share/ansible/collections` 
(https://github.com/ansible/ansible/blob/devel/lib/ansible/config/base.yml#L221) so the docs and behavior sound correct.

The only case that comes to mind that might be odd is if `~/.ansible` doesn't exist, but there are tons of other default config values that use ~/.ansible as well, so not a unique problem for collections_path.

**Comment by alikins** [comment]
Testing this out, it seems to do the right thing with a default config (or an ansible.cfg that doesn't provide `collections_path`).

`ansible-galaxy collection install alikins.collection_ntp` installs to ~/.ansible/collections/ansible_collections/alikins/collection_ntp as expected.

File: lib/ansible/cli/galaxy.py [hunk 5]
```diff
@@ -251,7 +251,8 @@ def add_install_options(self, parser, parents=None):
                                             "dependencies.".format(galaxy_type))
 
         if galaxy_type == 'collection':
-            install_parser.add_argument('-p', '--collections-path', dest='collections_path', required=True,
+            install_parser.add_argument('-p', '--collections-path', dest='collections_path',
+                                        default=C.COLLECTIONS_PATHS[0],
```
Comment (alikins) [inline lib/ansible/cli/galaxy.py:255]: The only odd behavior I can create is if I set 'collections_paths = :'  (an empty list)

``` ini
[defaults]
collections_paths = :
```

In that case, the collection ends up getting installed into $PWD/ansible_collections/.
But I suspect that is a general issue with config types of 'pathspec' and not specific
to collection_path.

**Review by alikins — APPROVED** [review]
LGTM.

All the normal use cases seems to work as expected.

**Comment by jborean93** [comment]
I’m going to leave now, the original decision was to not have a default because we couldn’t agree on what it should be. This needs to have a wider discussion before it is merged

**Comment by geerlingguy** [comment]
I agree that having the default be used would match current end user usage / expectations coming from the world of role installation.

`~/.ansible/collections` has my vote, if the idea is we're moving to a more user-centric view of Ansible content (vs. what I remember being the old default, of `/etc/ansible/*`).

**Comment by newswangerd** [comment]
We absolutely need to have a default path. I don't think we can expect people to know the correct directory to install collections to, especially if we want to convince people that collections don't significantly increase the complexity of managing your ansible installations.

**Review by bcoca — CHANGES_REQUESTED** [review]
we have this in roles, but we also check the path is 'writable' and choose first one, we should abstract into common fucntion for both to select the dir.

Also there is the issue when running as `root` you normally are trying to do 'system install' and should avoid ~/ entries

File: docs/docsite/rst/dev_guide/developing_collections.rst [hunk 6]
```diff
@@ -238,15 +238,19 @@ as a distribution method, but you can use it directly to install the collection
...
 -------------------------
 
-You can try your collection locally by installing it from the tarball.
+You can try your collection locally by installing it from the tarball. The following will enable an adjacent playbook to
+access the collection:
 
 .. code-block:: bash
 
-   ansible-galaxy collection install my_namespace-my_collection-1.0.0.tar.gz -p ./collections/ansible_collections
+   ansible-galaxy collection install my_namespace-my_collection-1.0.0.tar.gz -p ./collections
```
Comment (abadger) [inline docs/docsite/rst/dev_guide/developing_collections.rst:253]: @acozine @chouseknecht **If** this PR as a whole gets deferred to 2.10, should the above two changes be pushed into 2.9.x separately?  It seems like it fixes an independent documentation bug.

Comment (sivel) [reply docs/docsite/rst/dev_guide/developing_collections.rst:253]: I am pretty positive that the change to give a default install path will not be accepted for 2.9, potentially not even for 2.10.

File: lib/ansible/cli/galaxy.py (hunk 5 above)
Comment (nitzmahone) [reply lib/ansible/cli/galaxy.py:255]: Yeah, this should probably be an error if the config has explicitly removed the paths...

**Comment by chouseknecht** [comment]
Changes requested by @bcoca go beyond the scope of this PR, as it would require taking apart logic in the role install [found here]( https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/role.py#L294) and logic in the collection install [found here](https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/collection.py#L160). I'm not willing to tinker with all of that and take on the risk of breaking things.

Also, it's been suggest that C.COLLECTIONS_PATHS might not contain a value. I've tested this and found it to not be possible. If you don't define and env var, and you set `collections_paths` in ansible.cfg to "", it defaults to your home directory. If somehow you don't have a home dir, then maybe.

**Review by nitzmahone — COMMENTED** [review]
If we're going with this as-is, it just needs a changelog fragment, probably just a `minor_changes` section describing this update.

**Pull request merged**